vulnerabilities/import_runner.py

```diff
@@ -58,6 +61,46 @@ class ImportRunner:
     def __init__(self, importer: models.Importer, batch_size: int):
         self.importer = importer
         self.batch_size = batch_size
+        self._bulk_create_vuln_pkg_refs = []
+        self._bulk_update_vuln_pkg_refs = []
+        self._bulk_create_vuln_refs = []
+    
+    def __del__(self):
+        models.Vulnerability_Package_Relation.objects.bulk_create(self._bulk_create_vuln_pkg_refs, ignore_conflicts=True)
+        models.Vulnerability_Package_Relation.objects.bulk_update(self._bulk_update_vuln_pkg_refs,['is_vulnerable'])
+        models.VulnerabilityReference.objects.bulk_create(self._bulk_create_vuln_refs, ignore_conflicts=True)
+
+    # TODO: Currently the implementation may contain duplicate model instances in `_bulk_create_vuln_refs` and `_bulk_create_vuln_pkg_refs`
+    #       to make this work , `ignore_conflicts=True` is used. Find some way to remove duplicates of model instances(before writing them db)
+    #       and remove the   `ignore_conflicts=True` flag.
+    @property  
+    def bulk_create_vuln_pkg_refs(self):
+
+        if len(self._bulk_create_vuln_pkg_refs) >= self.batch_size : 
+            models.Vulnerability_Package_Relation.objects.bulk_create(self._bulk_create_vuln_pkg_refs, ignore_conflicts=True)
```

Comment:
Are you sure the performance gains are not mostly due to `ignore_conflicts=True`?

Reply:
No. The performance gain is due to removing the excess `get_or_create` and utilising the fact that, if 

A Vulnerability row didn't previously existed(created=True), then we for sure know that all objects which it is a foreign key in, don't exist too, hence use bulk create. Same goes with `Package`.

I had gotten it to work by simply having lists at a `Advisory` level for bulk inserts without ignore_conflicts, but that way sometimes I would bulk_create on a single item, or sometimes thousand items, everything depended on the nature Advisory being passed. 

The conflicts are caused due to duplicate model instances in. 
  ```
        self._bulk_create_vuln_pkg_refs = []
        self._bulk_update_vuln_pkg_refs = []
        self._bulk_create_vuln_refs = []

```
 I learned the hard way that I can't put these instances in a `set` because their hashes are created on basis of pk(which they get after being written to db) :) .